### PR TITLE
Sql injection

### PR DESCRIPTION
**Pull request by siddhantgosavi**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by siddhantgosavi** [comment]
Hi @bmoore-msft , could you please review the pull request.

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      }
+    },
+    "resources": [
+      {
+        "apiVersion": "2017-03-01-preview",
+        "type": "Microsoft.Sql/servers/auditingSettings",
+        "name": "[concat(parameters('sqlServerName'), '/', 'default')]",
+        "properties": {
+          "state": "Enabled",
+          "storageEndpoint": "[concat('https://',parameters('storageAccountName'),'.blob.core.windows.net')]",
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json]: use reference() to get the endpoint

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+          "storageAccountAccessKey": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName')), '2016-01-01').keys[0].value]",
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json]: update apiVersion

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+      "sqlServerName": {
+        "type": "string"
+      },
+      "storageAccountName": {
+        "type": "string",
+        "defaultValue": "storageaccountname"
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json]: remove this or use a uniqueString to generate a unique name

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    "resources": [
+      {
+        "apiVersion": "2015-05-01-preview",
+        "type": "Microsoft.Sql/servers/securityAlertPolicies",
+        "name": "[concat(parameters('sqlServerName'), '/', 'default')]",
+        "properties": {
+          "state": "Enabled",
+          "disabledAlerts": "",
+          "emailAddresses": "[parameters('sendAlertsTo')]",
+          "storageEndpoint": "[concat('https://',parameters('storageAccountName'),'.blob.core.windows.net')]",
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json]: use reference() to get the endpoint

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+          "storageAccountAccessKey": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', parameters('storageAccountName')), '2016-01-01').keys[0].value]",
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json]: update apiVersion

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    "parameters": {
+      "sqlServerName": {
+        "type": "string"
+      },
+      "location": {
+        "type": "string"
+      },
+      "administratorLogin": {
+        "type": "string",
+        "defaultValue": "sqladmin"
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json]: remove defaultValue - increases surface area

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      },
+      "tags": {
+        "type": "object"
+      }
+    },
+    "resources": [
+      {
+        "name": "[parameters('sqlServerName')]",
+        "type": "Microsoft.Sql/servers",
+        "apiVersion": "2015-05-01-preview",
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json]: update apiVersion

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      }
+    ],
+    "outputs": {
+      "name": {
+        "type": "string",
+        "value": "parameters('sqlServerName')"
+      },
+      "fqdn": {
+        "type": "string",
+        "value": "[reference(concat('Microsoft.Sql/servers/', parameters('sqlServerName')),'2015-05-01-preview').fullyQualifiedDomainName]"
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json]: update apiVersion

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+      },
+      "principalId": {
+        "type": "string",
+        "value": "[reference(concat('Microsoft.Sql/servers/', parameters('sqlServerName')),'2015-05-01-preview', 'Full').identity.principalId]"
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json]: update apiVersion

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.web/sites.json [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        "apiVersion": "2018-02-01",
+        "name": "[parameters('name')]",
+        "type": "Microsoft.Web/sites",
+        "tags": "[variables('combinedTags')]",
+        "identity": {
+          "type": "SystemAssigned"
+        },
+        "properties": {
+          "name": "[parameters('name')]",
+          "serverFarmId": "[concat('/subscriptions/', parameters('subscriptionId'),'/resourcegroups/', parameters('serverFarmResourceGroup'), '/providers/Microsoft.Web/serverfarms/', parameters('hostingPlanName'))]",
```
Comment (bmoore-msft) [inline 101-SQL-Injection-Attack-Prevention/nested/microsoft.web/sites.json]: simplify, just use resourceId()

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+      {
+        "name": "appgw-prevention",
+        "wafMode": "Prevention"
+      }
+    ],
+    "httpProbeName": "aseProbeHTTP",
+    "httpsProbeName": "aseProbeHTTPS",
+    "diagStorageAccName": "[concat('sqlinjection','stg',substring(uniqueString(resourceGroup().id),0,5))]",
+    "webAppName": "[concat('sql-injection-webapp','-',substring(uniqueString(resourceGroup().id),0,5))]",
+    "webAppUrl": "[concat(variables('webAppName'),'.azurewebsites.net')]",
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done. now getting webapp endpoint using reference().

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+    "aspName": "[concat('sql-injection-asp','-',substring(uniqueString(resourceGroup().id),0,5))]",
+    "appServiceConnectionType": "SQLAzure",
+    "sqlServerName": "[concat('sqlinjectionserver',substring(uniqueString(resourceGroup().id),0,5))]",
+    "databases": [
+      {
+        "name": "contosoclinic",
+        "edition": "Standard"
+      }
+    ],
+    "omsTemplateUri": "[concat(parameters('_artifactsLocation'),'/nested/microsoft.loganalytics/workspaces.json',parameters('_artifactsLocationSasToken'))]",
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done. using uri function to get the location. but still need to use concat to append the _artifactsLocationSasToken if it comes.

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.loganalytics/workspaces.json [hunk 13]
```diff
@@ -0,0 +1,115 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+      "omsWorkspaceName": {
+        "type": "string",
+        "defaultValue": "qwerty-workspace"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.loganalytics/workspaces.json]: yes it does. changing qwerty to oms and adding resource group unique string as well.

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json (hunk 1 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json]: done.

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json (hunk 2 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.auditingsettings.json]: done. 
2018-03-01-preview

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json (hunk 3 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json]: this needs to be a mandatory parameter. lets not assume any default value for it. i am removing it.

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json (hunk 4 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json]: done.

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json (hunk 5 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.securityalertpolicies.json]: done.

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json (hunk 7 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json]: found this one as the latest in the resource explorer
image

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json (hunk 8 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.sql/servers.v12.0.json]: found the same one as the latest

File: 101-SQL-Injection-Attack-Prevention/nested/microsoft.web/sites.json (hunk 10 above)
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/nested/microsoft.web/sites.json]: done.

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.parameters.json [hunk 14]
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "sqlAdministratorName": {
+      "value": "sqladmin"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.parameters.json]: done.

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,614 @@
...
+                "properties": {
+                  "RuleType": "Basic",
+                  "httpListener": {
+                    "id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/httpListeners/appGatewayHttpListener')]"
+                  },
+                  "backendAddressPool": {
+                    "id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/backendAddressPools/appGatewayBackendPool')]"
+                  },
+                  "backendHttpSettings": {
+                    "id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/backendHttpSettingsCollection/appGatewayBackendHttpSettings')]"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done.

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 16]
```diff
@@ -0,0 +1,614 @@
...
+            "value": [
+              {
+                "Name": "rule1",
+                "properties": {
+                  "RuleType": "Basic",
+                  "httpListener": {
+                    "id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/httpListeners/appGatewayHttpListener')]"
+                  },
+                  "backendAddressPool": {
+                    "id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/backendAddressPools/appGatewayBackendPool')]"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 17]
```diff
@@ -0,0 +1,614 @@
...
+            ]
+          },
+          "requestRoutingRules": {
+            "value": [
+              {
+                "Name": "rule1",
+                "properties": {
+                  "RuleType": "Basic",
+                  "httpListener": {
+                    "id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/httpListeners/appGatewayHttpListener')]"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 18]
```diff
@@ -0,0 +1,614 @@
...
+          "httpListeners": {
+            "value": [
+              {
+                "name": "appGatewayHttpListener",
+                "properties": {
+                  "FrontendIPConfiguration": {
+                    "Id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/frontendIPConfigurations/appGatewayFrontendIP')]"
+                  },
+                  "FrontendPort": {
+                    "Id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/frontendPorts/appGatewayFrontendPort')]"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 19]
```diff
@@ -0,0 +1,614 @@
...
+              }
+            ]
+          },
+          "httpListeners": {
+            "value": [
+              {
+                "name": "appGatewayHttpListener",
+                "properties": {
+                  "FrontendIPConfiguration": {
+                    "Id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name), '/frontendIPConfigurations/appGatewayFrontendIP')]"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done

File: 101-SQL-Injection-Attack-Prevention/azuredeploy.json [hunk 20]
```diff
@@ -0,0 +1,614 @@
...
+            "value": [
+              {
+                "name": "appGatewayBackendHttpSettings",
+                "properties": {
+                  "Port": 80,
+                  "Protocol": "Http",
+                  "CookieBasedAffinity": "Disabled",
+                  "pickHostNameFromBackendAddress": "true",
+                  "Probe": {
+                    "id": "[concat(resourceId('Microsoft.Network/applicationGateways',variables('applicationGateways')[copyIndex()].name),'/probes/',variables('httpProbeName'))]"
```
[the remark this answers is not in the document]
Comment (siddhantgosavi) [reply 101-SQL-Injection-Attack-Prevention/azuredeploy.json]: done

**Comment by bmoore-msft** [comment]
@siddhantgosavi changes look good, can you check the merge conflicts?

**Comment by siddhantgosavi** [comment]
Hi @bmoore-msft , I've raised a separate PR including both sql injection and xss scenario. This one can be closed.

**Pull request closed without merging**